### DISCUSSION TITLE
it needs double "C-g" to quit corfu completion in org-mode buffer

### DISCUSSION THREAD

**Discussion by RNAer** [minad/corfu#541] [Q&A]
Normally, I can type C-g to run `corfu-quit` to exit the completion. Weirdly, in org-mode buffer, I need to type it twice. Trying to figure it out why, I added the following hook to see what's happening when I run `C-g`.

```
(add-hook 'post-command-hook
          (lambda ()
            (message "Command executed: %s" this-command)))
```

I find that the first `C-g` is calling `keyboard-quit` instead of `corfu-quit` (see the crude recording)  and this only occurs in org-mode buffer. I am wondering if anyone knows why and how to fix it.

https://github.com/user-attachments/assets/b02f3d2a-dcd5-4677-a5d9-40c1677bf1d1

I am using GNU Emacs 30.0.93 (build 2, aarch64-apple-darwin24.2.0, NS appkit-2575.30 Version 15.2 (Build 24C101)) and latest corfu and builtin org-mode. Note that this problem persists even I purge all my org-mode settings.

**Comment by minad** [comment]
I don't see this problem. Can you reproduce this from emacs -Q?

**Reply by RNAer** [reply]
I tried that and found the culprit - it is due to the jupyter, especially the `org-babel-do-load-languages` setting below. Delete those 3 lines and C-g works as expected. Not sure what's wrong with those lines...

```
  (use-package jupyter
    :commands
    (jupyter-run-server-repl
     jupyter-run-repl
     jupyter-server-list-kernels)
    :custom-face
    (jupyter-repl-traceback ((t (:background "gray9"))))
    :init
    (org-babel-do-load-languages 'org-babel-load-languages
                                 (append org-babel-load-languages
                                         '((jupyter     . t))))
    (setq jupyter-repl-echo-eval-p t)
    (setq jupyter-repl-allow-RET-when-busy t)
    (setq org-babel-default-header-args:jupyter-python
          '((:async . "yes")
            (:results . "replace")
            (:session . "py"))))
```

**Reply by kayzhang** [reply]
Got the same/similar issue where C-g cannot trigger corfu-quit immediately after the completion popup is shown. I need to either 1) "C-g C-g" or 2) wait for a bit (like 2 seconds) and then C-g will work.

And found that it was caused by other packages.
* In prog-mode, if I remove [indent-bars](https://github.com/jdtsmith/indent-bars) from my config, then the issue will be resolved.
* In org-mode, if I remove [org-modern-indent](https://github.com/jdtsmith/org-modern-indent) from my config, then the issue will be resolved.

I'm not gonna dive deep into this since indent-bars and org-modern-indent are not a must have for me. So, I'll just remove them to workaround this for now.

**Reply by jdtsmith** [reply]
Maybe `M-x toggle-debug-on-quit` and report the backtrace of the first "failed" C-g.

**Reply by kayzhang** [reply]
@jdtsmith Hi, seems like there was no backtrace reported. The first C-g (if clicked immediately after the Corfu popup is shown) triggered the visual-bell (flashing my modeline) and logged "Quit" in the left bottom corner.

I debugged with [org-modern-indent](https://github.com/jdtsmith/org-modern-indent) since it's a small package and narrowed down the issue to the call [here](https://github.com/jdtsmith/org-modern-indent/blob/main/org-modern-indent.el#L357-L358) and the flyspell mode (I used `brew install ispell` for the installation).

My setup:

> macOS: 15.4.1
> emacs: 30.1 (installed by `brew install --cask emacs`)
> Corfu: 2.1
> org-modern-indent (HEAD at [9973bd3b91e4733a3edd1fca232208c837c05473](https://github.com/jdtsmith/org-modern-indent/commit/9973bd3b91e4733a3edd1fca232208c837c05473))
> ispell: 3.4.06 ((installed by `brew install ispell`)

The repro test.el:

```
(require 'package)
(package-initialize)
(setq inhibit-startup-screen t)
(use-package corfu
  :demand t
  :custom
  (corfu-auto t)
  :init
  (global-corfu-mode))

(setq char-property-alias-alist '((face font-lock-face) (display blabla)))

(flyspell-mode)
```

Detailed explanation of the repro steps:

#### 1. Load Corfu with `emacs -q -l test.el` with the following test.el (No-Issue until now):

```
(require 'package)
(package-initialize)
(setq inhibit-startup-screen t)
(use-package corfu
  :demand t
  :custom
  (corfu-auto t)
  :init
  (global-corfu-mode))
```

Now the value of `char-property-alias-alis` is `((face font-lock-face))` which seems to be set by Corfu itself [here](https://github.com/minad/corfu/blob/10e24c8bdbdd4e6d3145878f3ce4357c3753b0a1/extensions/corfu-popupinfo.el#L133).

Now, `corfu-quit` works as expcted for the first C-g.

#### 2. Change the value of `char-property-alias-alis` (No-Issue until now)

We can do either of the two options below:
* Option 1. Install org-modern-indent and then enable org-modern-indent-mode which will change `char-property-alias-alis` to `((display org-modern-indent-display) (face font-lock-face))`.
* Option 2. Just manually change the value of `char-property-alias-alis` by `(setq char-property-alias-alist '((face font-lock-face) (display blabla)))`.

Now, `corfu-quit` still works as expcted for the first C-g.

#### 3. Now enable `flyspell-mode` (Now Issue is introduced)

Run `M-x flyspell-mode`. Then the issue is introduced!

This can be recovered by either:
* Option 1 - run `M-x flyspell-mode` again to disable ispell
* Option 2 - run `(setq char-property-alias-alist '((face font-lock-face)))` to recover the `char-property-alias-alist`.

Hope this will help.

Thanks,
Kai

**Reply by kayzhang** [reply]
Also, seems like this issue is specific to C-g instead of `corfu-quit`. For example, if I bind C-f to `corfu-quit` with:

```
(require 'package)
(package-initialize)
(setq inhibit-startup-screen t)
(use-package corfu
  :demand t
  :custom
  (corfu-auto t)
  :init
  (global-corfu-mode)
  :bind
  (:map corfu-map
    ("C-f" . #'corfu-quit)))
```

Then,
* The first C-f immediately after Corfu popup is shown will trigger `corfu-quit` correctly as expected.
* The first C-g immediately after Corfu popup is shown still have the same issue - it will still trigger the visual-bell (flashing my modeline) and log "Quit" in the left bottom corner.

**Reply by jdtsmith** [reply]
I don't have a great theory, but `C-g` is globally bound to `keyboard-quit`.  Only in `corfu-map` is it re-bound to `corfu-quit`.  Perhaps `flyspell` is stealing focus from corfu's child frame, such that its keymap is not active?  Presumably this happens in its post-command-hook, where it uses `with-local-quit`, which appears to "post" another quit to `quit-flag`.  I don't understand how the `display` alias matters.  `indent-bars` also uses a display alias.  If you re-bind to another key, does `corfu-quit` always work as expected?

**Reply by kayzhang** [reply]
> I don't understand how the display alias matters. indent-bars also uses a display alias.

As I mentioned earlier, `indent-bars` caused the same issue for me.

> If you re-bind to another key, does corfu-quit always work as expected?

Yes. If I re-bind `corfu-quit` to C-f, then C-f will always work. But the first C-g still does not work even if I re-bind C-g to something else in the corfu-map.

Some more details to clarify.
1. The issue only occurs to me when:
   * `flyspell-mode` is enabled (Note, `ispell-minor-mode` won't repro this issue)
   * AND `char-property-alias-alist` has some extra alias besides the `((face font-lock-face))` which is set by Corfu. Actually whether it's `display` or not does not matter. For example, set it to `((face font-lock-face) (foo bar))` will also repro this issue.
2. The issue is not specific to `corfu-quit` but instead is specific to C-g.
   * Re-bind `corfu-quit` to other key-binding like C-f. Then C-f will always trigger `corfu-quit` as expected.
3. It seems not related to the global keybinding of C-g to `keyboard-quit` as well. For example, if I do both of 1) bind C-g in corfu-map to `find-file` AND 2) `(global-set-key (kbd "C-g") 'find-file)`, then in theory C-g is not bound to either `corfu-quit` or `keyboard-quit`. However, the actual behavior is as follows.
   * With 1) flyspell-mode enabled AND 2) `char-property-alias-alist` set to `((face font-lock-face) (foo bar))`, then the first C-g will still trigger the visual-bell.
4. As I mentioned previously, the issue was only reproducible for the first C-g immediately after the Corfu popup is shown. If I wait for like 2 or 3 seconds after the popuo is shown, then the first C-g after that won't trigger this issue.

Based on the above, it seems like that in the first 2 or 3 seconds after Corfu popup is shown, flyspell-mode 1) re-binds C-g to `keyboard-quit` no matter whether C-g is bound to something else like `find-file`, and 2) it steals focus from Corfu.

**Reply by kayzhang** [reply]
Looked into the [flyspell.el](https://github.com/emacs-mirror/emacs/blob/master/lisp/textmodes/flyspell.el) and if I remove the `with-local-quit` [here](https://github.com/emacs-mirror/emacs/blob/master/lisp/textmodes/flyspell.el#L968) in the `flyspell-post-command-hook` function then the issue will disappear.

Added some logs into the `flyspell-post-command-hook` function, and the 2 or 3 second immediately after the Corfu popup is shown is exactly the time when the flyspell executes the `flyspell-post-command-hook` method. Then I think the root cause is that during that 2 or 3 seconds, C-g is treated as quitting the `flyspell-post-command-hook` function.

**Reply by jdtsmith** [reply]
Try evaluating this:

```elisp
(trace-function #'keyboard-quit nil
		(lambda ()
		  (concat (format  "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook)
			  (backtrace-to-string (backtrace-get-frames 'backtrace)))))
```

Then set things up (no `debug-on-quit` needed) and try your failing `C-g`.  You should see the active buffer, post command hook, and backtrace in the `*trace-output*` buffer.  

One issue is that flyspell mode should not be active in the ` *corfu*` buffer.  I wonder if it is somehow being activated.

**Reply by jdtsmith** [reply]
@kayzhang Did your logs indicate in which buffer `flyspell` was running its `post-command-hook`?

**Reply by kayzhang** [reply]
With

```
(trace-function #'keyboard-quit nil
		(lambda ()
		  (concat (format  "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook)
			  (backtrace-to-string (backtrace-get-frames 'backtrace)))))
```

The werid behavior is that if I just typed C-g to trigger keyboard-quit in the scratch buffer, it will log the following logs:

```
======================================================================
1 -> (keyboard-quit)
>> In buffer *scratch*:
PCH: (jit-lock--antiblink-post-command eldoc-schedule-timer evil--jump-handle-buffer-crossing corfu--auto-post-command yas--post-command-handler t flyspell-post-command-hook)
  backtrace-to-string(nil)
  (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))
  #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace))))()
  trace--entry-message(keyboard-quit 1 nil #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))))
  #f(compiled-function (body &rest args) #<bytecode -0xd04c22d7ecb7778>)(#f(advice ad-Advice-keyboard-quit :around #f(compiled-function () (interactive nil) #<bytecode 0xe851acf88016143>)))
  apply(#f(compiled-function (body &rest args) #<bytecode -0xd04c22d7ecb7778>) #f(advice ad-Advice-keyboard-quit :around #f(compiled-function () (interactive nil) #<bytecode 0xe851acf88016143>)) nil)
  keyboard-quit()
  funcall-interactively(keyboard-quit)
  call-interactively(keyboard-quit nil nil)
  command-execute(keyboard-quit)

1 <- keyboard-quit: !non-local\ exit!
>> In buffer *scratch*:
PCH: (jit-lock--antiblink-post-command eldoc-schedule-timer evil--jump-handle-buffer-crossing corfu--auto-post-command yas--post-command-handler t flyspell-post-command-hook)
  backtrace-to-string(nil)
  (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))
  #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace))))()
  trace--exit-message(keyboard-quit 1 !non-local\ exit! #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))))
  #f(compiled-function () #<bytecode 0x12bcfa14339d7422>)()
  #f(compiled-function (body &rest args) #<bytecode -0xd04c22d7ecb7778>)(#f(advice ad-Advice-keyboard-quit :around #f(compiled-function () (interactive nil) #<bytecode 0xe851acf88016143>)))
  apply(#f(compiled-function (body &rest args) #<bytecode -0xd04c22d7ecb7778>) #f(advice ad-Advice-keyboard-quit :around #f(compiled-function () (interactive nil) #<bytecode 0xe851acf88016143>)) nil)
  keyboard-quit()
  funcall-interactively(keyboard-quit)
  call-interactively(keyboard-quit nil nil)
  command-execute(keyboard-quit)
```

**However, the first C-g in the corfu popup did not trigger any traces added above.**

I then added the similar log with the code snippet you just shared for the `flyspell-post-command-hook` function. And the log for the `flyspell-post-command-hook` function was as follows:

```
======================================================================
1 -> (flyspell-post-command-hook)
>> In buffer *scratch*:
PCH: (jit-lock--antiblink-post-command eldoc-schedule-timer evil--jump-handle-buffer-crossing corfu--auto-post-command yas--post-command-handler t flyspell-post-command-hook)
  backtrace-to-string(nil)
  (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))
  #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace))))()
  trace--entry-message(flyspell-post-command-hook 1 nil #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))))
  #f(compiled-function (body &rest args) #<bytecode -0xd02169ed1eb7778>)(#f(lambda () [message-signature-separator flyspell-mode-map t] "The `post-command-hook' used by flyspell to check a word on-the-fly." (interactive nil) (if flyspell-mode (progn (condition-case nil (let ((inhibit-quit nil)) (let (... deactivate-mark) (if ... ...) (if ... ...) (if ... ... ... ... ...) (while ... ...) (setq flyspell-previous-command command))) (quit (setq quit-flag t) (eval '... t)))))))
  apply(#f(compiled-function (body &rest args) #<bytecode -0xd02169ed1eb7778>) #f(lambda () [message-signature-separator flyspell-mode-map t] "The `post-command-hook' used by flyspell to check a word on-the-fly." (interactive nil) (if flyspell-mode (progn (condition-case nil (let ((inhibit-quit nil)) (let (... deactivate-mark) (if ... ...) (if ... ...) (if ... ... ... ... ...) (while ... ...) (setq flyspell-previous-command command))) (quit (setq quit-flag t) (eval '... t)))))) nil)
  flyspell-post-command-hook()

1 <- flyspell-post-command-hook: self-insert-command
>> In buffer *scratch*:
PCH: (jit-lock--antiblink-post-command eldoc-schedule-timer evil--jump-handle-buffer-crossing corfu--auto-post-command yas--post-command-handler t flyspell-post-command-hook)
  backtrace-to-string(nil)
  (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))
  #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace))))()
  trace--exit-message(flyspell-post-command-hook 1 self-insert-command #f(lambda () [t] (concat (format "\n>> In buffer %s:\nPCH: %S\n" (buffer-name) post-command-hook) (backtrace-to-string (backtrace-get-frames 'backtrace)))))
  #f(compiled-function () #<bytecode 0x12bcc5dd2f5f7422>)()
  #f(compiled-function (body &rest args) #<bytecode -0xd02169ed1eb7778>)(#f(lambda () [message-signature-separator flyspell-mode-map t] "The `post-command-hook' used by flyspell to check a word on-the-fly." (interactive nil) (if flyspell-mode (progn (condition-case nil (let ((inhibit-quit nil)) (let (... deactivate-mark) (if ... ...) (if ... ...) (if ... ... ... ... ...) (while ... ...) (setq flyspell-previous-command command))) (quit (setq quit-flag t) (eval '... t)))))))
  apply(#f(compiled-function (body &rest args) #<bytecode -0xd02169ed1eb7778>) #f(lambda () [message-signature-separator flyspell-mode-map t] "The `post-command-hook' used by flyspell to check a word on-the-fly." (interactive nil) (if flyspell-mode (progn (condition-case nil (let ((inhibit-quit nil)) (let (... deactivate-mark) (if ... ...) (if ... ...) (if ... ... ... ... ...) (while ... ...) (setq flyspell-previous-command command))) (quit (setq quit-flag t) (eval '... t)))))) nil)
  flyspell-post-command-hook()
```

There were a huge lots of such logs for `flyspell-post-command-hook` and there was no patterns like `In buffer *corfu*:`.

**Reply by jdtsmith** [reply]
> the first C-g in the corfu popup did not trigger any traces added above.

Well that's good.  You want `corfu-quit` to be called, not `keyboard-quit`.  Is it?  You could `M-x trace-function corfu-quit` and see it logged.

**Reply by jdtsmith** [reply]
Also, it seems `C-g` handling is quite complex, and (sometimes) operates on a deeper level than ordinary keymaps:

>    Typing ‘C-g’ while the command loop is waiting for keyboard input does not cause a quit; it acts as an ordinary input character.  In the simplest case, you cannot tell the difference, because ‘C-g’ normally runs the command ‘keyboard-quit’, whose effect is to quit.

The corollary is that when the command loop is _not_ waiting for input (e.g. because flyspell is keeping it busy), it does _not_ act as an ordinary input character, so the active keymap isn't consulted.   You could confirm this by using `set-input-mode` to set `QUIT-CHAR` to another key, and see whether the problem disappears.

**Reply by kayzhang** [reply]
@jdtsmith I tried the following.

1. `M-x trace-function` to trace `keyboard-quit` and `corfu-quit`. And the first failed C-g did not trigger any of them. I think this aligns with your corollary - it triggered a "quit" instead of "keyboard-quit"/"corfu-quit" since the command loop is NOT waiting for input.
2. I tried `(set-input-mode t nil 0 20)` where 20 is the ASCII for C-t (not sure why it does not work in graphical emacs but only work in terminal emacs, but anyway). The original value of `current-input-mode` is `t nil 0 7` where 7 is the ASCII for C-g.
   * After setting "quit" to C-t. Then the first C-g triggers corfu-quit as expected.
   * But now the first C-t triggers the original issue.

I think all of the above observations aligns with your corollary: 

>when the command loop is not waiting for input (e.g. because flyspell is keeping it busy), it does not act as an ordinary input character, so the active keymap isn't consulted

Or more specifically, the first C-g (or C-t after `(set-input-mode t nil 0 20)`) triggers "quit" actually.

One callout is that due to some weird reasons this issue is somehow related to the extra alias in `char-property-alias-alist`.

**Reply by kayzhang** [reply]
@jdtsmith So, now we have sort of found the root cause. Personally, I do not think this issue needs to be prioritized. I could just re-bind `corfu-quit` to another keybinding. Will let you to make the decision here. Thanks.

**Reply by jdtsmith** [reply]
Yeah, probably flyspell should return quicker and work asynchronously.

**Reply by minad** [reply]
The inefficiency of Flyspell was one of the motivations for [Jinx](https://github.com/minad/jinx).

**Reply by kayzhang** [reply]
@minad Sure. I'll definitely give Jinx a try later.

**Reply by ISouthRain** [reply]
Hello? 
Is there any good solution to this problem?
Does this look like a flyspell-mode issue?
I disabled flyspell-mode OK, it really fixed the problem.
Is there a better solution?

**Reply by ISouthRain** [reply]
(setq flyspell-delay-use-timer t) ;; fix need double "C-g" to quit corfu

**Reply by shahinism** [reply]
For me disabling `flyspell` didn't solve the issue, so probably there is something else on its way. Same behavior, logging `Quit` and flashing modeline. For the time being, I'm just binding it to `<escape>`.

**Comment by hanwenguo** [comment]
Sorry if disturbing anyone, but I have encountered a similar yet different issue. The behavior seems similar: if I press `C-g` at some point (being too early or too late won't trigger the issue) after triggering corfu popup, that `C-g` won't work.

However, the differences are:
- With the messaging setup from OP, the logged command for that `C-g` is `nil` instead of `keyboard-quit`.
- After `(set-input-mode t nil 0 20)`, the same issue with `C-g` still happens. (`C-t` will always trigger command `nil` then, but that is natural since `C-t` is not bound in `corfu-map`.)

Although coming up with a guess is generally bad when reporting bugs, but for this one, my guess is that it's somehow the inverse of the above discussion: the command loop _is_ waiting for keyboard input and `C-g` thus acts as an ordinary input character, which is mapped to `nil` in some weird keymap activated for some weird reason.

Would appreciate any suggestion on debugging.

**Reply by minad** [reply]
Do you even observe this in a minimal setup starting from emacs -Q with only Corfu loaded, or only few other modes?

**Reply by hanwenguo** [reply]
Yes. It seems that this is some issue between Prot's [spacious-padding](https://github.com/protesilaos/spacious-padding) and `corfu-popupinfo-mode`. Below is a minimal example with `emacs -Q`:

```emacs-lisp
(use-package spacious-padding)

(use-package corfu
  :config
  (setq corfu-auto t
        corfu-auto-prefix 1 ;; just to help with debugging
  )
  (corfu-popupinfo-mode 1)
  (setq corfu-popupinfo-delay 1.0) ;; just to help with debugging
)

(global-corfu-mode 1)
(spacious-padding-mode 1)
```

To trigger, type anything and try to press `C-g` just before (or after? I'm not sure as the time frame is very short) the moment the info popup is about to appear. 

Relevant or not, I'm on Emacs 31.0.50 feature/igc and corfu 9d5fc8f. ~~I'll test with HEAD corfu again.~~ EDIT: still happening on current HEAD (79bb30e)
